typed-protocols/src/Network/TypedProtocol/Driver.hs

```diff
@@ -129,6 +137,16 @@ runPeer tr Codec{encode, decode} channel@Channel{send} =
         Left failure ->
           throwM failure
 
+    -- This cases is a protocol error: the remote peer has sent further data
+    -- after the final message.
+    go (Just _) (Done _ _) = throwM IllegalTrailingData
```

Comment:
This doesn't seem right. It's legitimate for the channel to contain extra data, even if the peer is finished. Take the case of running the handshake protocol prior to some protocol determined by it: handshake may finish normally, leaving extra bytes that were sent according to the next protocol. In fact, that's a problem even without this change... we need to push the trailing data back onto the channel, so that it can be used to run the next protocol.

Reply:
Yes, fair enough. I agree with that.

> In fact, that's a problem even without this change... we need to push the trailing data back onto the channel, so that it can be used to run the next protocol.

Yes, or `runPeer` can return it.

Reply:
Each protocol, has it's own queue of received bytes, which means that trailing bytes of various protocols will not mix up.  With handshake it's slightly different, and we can solve this by moving handshake to run using mux layer, rather than handshake being run before mux.

Reply:
To be more precise: in PR 584, hanshake messages are wrapped using `MuxSDU` and are read / written using `MuxBearer`.  It has the property that it reads a single `MuxSDU` at a time, so any bytes sent eagerly will await in the kernel.

Reply:
> It's legitimate for the channel to contain extra data, even if the peer is finished.

This is the difference between a stream based and record based transport.  I would assume there should be no additional data within the record and all protocol messages should start at a record boundary.

Reply:
> Each protocol, has it's own queue of received bytes, which means that trailing bytes of various protocols will not mix up. With handshake it's slightly different, and we can solve this by moving handshake to run using mux layer, rather than handshake being run before mux.

You're speaking about very particular applications of typed-protocols. I'm speaking about how it ought to be in general. Your proposal is to do something rather weird and unexpected just because it's enough (or maybe even helpful) for the ouroboros-network mux mini-protocol stuff. I'm saying no, let's have it do the more permissive and general thing, because it's quite easy to get the "no-trailing-bytes" rule after-the-fact.

Reply:
For analogy: you wouldn't define a CBOR decoder to always fail if there are extra bytes.

Reply:
Instead, you let it return trailing bytes, and _if it's right for your application_ you can fail in case there are any.

Reply:
I wasn't seaking about `type-protocols`, we indeed ought to do that right; I'm just commenting that `ouroboros-network` will work fine.